Add `-n, --names` option to specify space names to export

Yousef-Majidi opened the pull request:
I've added a new option: `--names` and added it to the `ExportSpaces` method. 
When provided, it is a comma separated string, but the values are trimmed. So, both `"test1,test2"` and `"test1, test2"` will work. 

Let me know if you need any changes.

closes #4

frankkilcommins commented on src/Explore.Cli/Program.cs:
```diff
@@ -27,6 +27,15 @@ public static async Task<int> Main(string[] args)
+        var exportPath = new Option<string>(name: "--export-path", description: "The path for exporting files. It can be either a relative or absolute path") { IsRequired = false };
+        exportPath.AddAlias("-ep");
+
+        var exportFileName = new Option<string>(name: "--export-name", description: "The name of the file to export") { IsRequired = false };
```
Perhaps we could just use `file-path` also for this purpose and just enhance the description such that it covers both?

Yousef-Majidi replied:
I originally considered using file-path. However, file-path is marked as required. Did you want the export path to be required or did you want me to change it so that file-path is not required? In my mind, it didn't make sense to have file-path not required as it is used for importing. But also, it didn't make sense to force the user to always provide an export path

frankkilcommins replied:
I'd just create a new option variable for the export use case, and have the Option.name = file-path, and isRequired = false.

Example:
```
var exportFilePath = new Option<string>(name: "--file-path", description: "The path to the file used for importing data") { IsRequired = false };
filePath.AddAlias("-fp");
```

frankkilcommins commented on src/Explore.Cli/Program.cs:
```diff
@@ -348,10 +366,36 @@ internal static async Task ExportSpaces(string exploreCookie, bool? verboseOutpu
             };
 
+            // set the file name if provided
+            var fileName = string.IsNullOrEmpty(exportFileName) ? "ExploreSpaces.json" : $"{exportFileName}.json";
```
We should handle the scenario where folks might append `.json` to the fileName and also error if they append an alternative file extension.

frankkilcommins commented on src/Explore.Cli/Program.cs:
```diff
@@ -348,10 +366,36 @@ internal static async Task ExportSpaces(string exploreCookie, bool? verboseOutpu
+                // combine the path and filename
+                filePath = Path.Combine(exportPath, fileName);
+
+                if (!Directory.Exists(exportPath))
```
We should also check/handle permissions issues where we may not be able to create the directory or write the file to the specified directory. See the importSpaces function for similar logic

frankkilcommins commented:
Thanks for this......added a few comments above

Yousef-Majidi commented:
Hi @frankkilcommins. I've made the requested changes. 
I added a check for valid file extensions. I also added a check for the `\` or `/` character in the file name. 
I handled the permission checks. 
I also modified the exportPath variable. Using `-fp, --file-path` now. 

Let me know if you need more changes.

frankkilcommins commented on README.md:
```diff
@@ -131,6 +131,9 @@ From SwaggerHub Explore, navigate to your browser development tools, locate the
 **Options:**
   > -ec, --explore-cookie <explore-cookie> (REQUIRED)  A valid and active SwaggerHub Explore session cookie
+  > -ep, --export-path <export-path>                   The path to the directory used for exporting data. It can be either relative or absolute
+  > -en, --export-name <export-name>                   The name of the exported file
```
ReadME will need to be updated to reflect the changes

frankkilcommins commented on src/Explore.Cli/Program.cs:
```diff
@@ -252,17 +261,61 @@ internal static async Task ExportSpaces(string exploreCookie, bool? verboseOutpu
+                {
+                    fileName = $"{exportFileName}.json";
+                }
+                else if (exportFileName.EndsWith(".json"))
```
Maybe change to `EndsWith(".json", StringComparison.OrdinalIgnoreCase)` to support case insensitivity

frankkilcommins commented on src/Explore.Cli/Program.cs:
```diff
@@ -252,17 +261,61 @@ internal static async Task ExportSpaces(string exploreCookie, bool? verboseOutpu
+                }
+
+                // check for invalid characters in the file name
+                if (exportFileName.Contains('\\') || exportFileName.Contains('/'))
```
This is probably not an exhaustive check of invalid chars. It might be more robust to add some methods to UtilityHelper.cs to allow for robust checking.

Example:
```
    private static bool IsValidFileName(string fileName)
    {
        return (fileName.IndexOfAny(Path.GetInvalidFileNameChars()) >= 0);
    }

    private static bool IsValidFilePath(string filePath)
    {
        return (filePath.IndexOfAny(Path.GetInvalidPathChars()) >= 0);
    }
```

frankkilcommins replied:
Please add some UtilityHelperTests for above too 🙏

frankkilcommins commented:
Thanks for all the work. Nearly there I reckon.
Could you also supply some information on how you tested/verified the changes?

Yousef-Majidi commented:
- updated readme
- added validation methods for file-path and file-name
- moved file-path and file-name validation higher up in the code to prevent the panel from printing in invalid cases
- added error handling for permission denied cases for directory creation and also writing to file system
- added validation tests for `UtilityHelper`

Yousef-Majidi commented:
> Thanks for all the work. Nearly there I reckon. Could you also supply some information on how you tested/verified the changes?

I'm testing by manually building the app and testing it with my own account. All of the new commands added are easily verifiable by checking the output after each test. 

The validation checks are tested with the unit tester, but I realize some of them failed. I will have a look at this in the next couple of days. It's an OS dependant problem I think. Because `GetInvalidChars` is not guaranteed to contain a complete set of characters that are invalid.

Yousef-Majidi commented:
Hi @frankkilcommins. Sorry for the delay in fixing this. But I came up with somewhat of a brute force solution for the failed tests. Please review and let me know if you have a better solution. I think I already covered everything else that was requested. But if I missed anything, please let me know.

frankkilcommins reviewed (approved):
LGTM
